File: src/main/java/org/jetbrains/pmdkotlin/lang/kotlin/rule/KotlinRuleViolationFactory.java
package org.jetbrains.pmdkotlin.lang.kotlin.rule;

import net.sourceforge.pmd.Rule;
import net.sourceforge.pmd.RuleContext;
import net.sourceforge.pmd.RuleViolation;
import net.sourceforge.pmd.lang.ast.Node;
import net.sourceforge.pmd.lang.rule.AbstractRuleViolationFactory;
import net.sourceforge.pmd.lang.rule.ParametricRuleViolation;
import net.sourceforge.pmd.lang.rule.RuleViolationFactory;
import org.jetbrains.pmdkotlin.lang.kotlin.ast.AbstractKotlinNode;

public class KotlinRuleViolationFactory  extends AbstractRuleViolationFactory{
    public static final RuleViolationFactory INSTANCE = new KotlinRuleViolationFactory();

    @Override
    protected RuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message) {
        return new ParametricRuleViolation<AbstractKotlinNode>(rule, ruleContext, (AbstractKotlinNode) node, message);
    }

    @Override
    protected RuleViolation createRuleViolation(Rule rule, RuleContext ruleContext, Node node, String message, int beginLine, int endLine) {
        final ParametricRuleViolation<AbstractKotlinNode> violation = new ParametricRuleViolation<AbstractKotlinNode>(rule,
                ruleContext, (AbstractKotlinNode) node, message);
        violation.setLines(beginLine, endLine);
        return violation;
    }
}

File: src/main/kotlin/org/jetbrains/pmdkotlin/lang/kotlin/rule/KotlinRuleViolationFactory.kt
package org.jetbrains.pmdkotlin.lang.kotlin.rule

import net.sourceforge.pmd.Rule
import net.sourceforge.pmd.RuleContext
import net.sourceforge.pmd.RuleViolation
import net.sourceforge.pmd.lang.ast.Node
import net.sourceforge.pmd.lang.rule.AbstractRuleViolationFactory
import net.sourceforge.pmd.lang.rule.ParametricRuleViolation
import net.sourceforge.pmd.lang.rule.RuleViolationFactory
import org.jetbrains.pmdkotlin.lang.kotlin.ast.AbstractKotlinNode

class KotlinRuleViolationFactory : AbstractRuleViolationFactory() {

    override fun createRuleViolation(rule: Rule, ruleContext: RuleContext, node: Node, message: String): RuleViolation {
        return ParametricRuleViolation(rule, ruleContext, node as AbstractKotlinNode, message)
    }

    override fun createRuleViolation(rule: Rule, ruleContext: RuleContext, node: Node, message: String, beginLine: Int, endLine: Int): RuleViolation {
        val violation = ParametricRuleViolation(rule,
                ruleContext, node as AbstractKotlinNode, message)
        violation.setLines(beginLine, endLine)
        return violation
    }

    companion object {
        val INSTANCE: RuleViolationFactory = KotlinRuleViolationFactory()
    }
}
